sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -0,0 +1,231 @@
+            PartitionManager = partitionManager;
+            Options = options?.Clone() ?? new EventProcessorOptions();
+
+            Identifier = Guid.NewGuid().ToString();
+            PartitionPumps = new ConcurrentDictionary<string, PartitionPump>();
+        }
+
+        /// <summary>
+        ///   Starts the event processor.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task StartAsync()
+        {
+            if (RunningTask == null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask == null)
+                    {
+                        RunningTaskTokenSource = new CancellationTokenSource();
+
+                        PartitionPumps.Clear();
+
+                        var partitionIds = await InnerClient.GetPartitionIdsAsync().ConfigureAwait(false);
+
+                        await Task.WhenAll(partitionIds
+                            .Select(partitionId =>
+                            {
+                                var partitionContext = new PartitionContext(InnerClient.EventHubPath, ConsumerGroup, partitionId);
+                                var checkpointManager = new CheckpointManager(partitionContext, PartitionManager, Identifier);
+
+                                var partitionProcessor = PartitionProcessorFactory.CreatePartitionProcessor(partitionContext, checkpointManager);
+
+                                var partitionPump = new PartitionPump(InnerClient, ConsumerGroup, partitionId, partitionProcessor, Options);
+                                PartitionPumps.TryAdd(partitionId, partitionPump);
+
+                                return partitionPump.StartAsync();
+                            })).ConfigureAwait(false);
+
+                        RunningTask = RunAsync(RunningTaskTokenSource.Token);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Stops the event processor.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task StopAsync()
+        {
+            if (RunningTask != null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask != null)
+                    {
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        await RunningTask.ConfigureAwait(false);
```

Comment:
You may want to either set the task to null before awaiting (using an intermediate variable) or wrap in a try/finally where you null out the things to manage state... just in case your task code throws in the `await` so that you're not left in an inconsistent state.